concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1155 @@
+)
+WithServiceBefore(host, svcName, apiRule) // this applies a side effect
+```
+</details>
+
+<details>
+	<summary>Do</summary>
+
+In this example `WithServiceAfter` is used as an argument to `NewAPIRule`.
+
+```go
+func WithServiceAfter(name, host string) APIRuleOption {
+	return func(r *apigatewayv1alpha1.APIRule) {
+		port := uint32(443)
+		isExternal := true
+		r.Spec.Service = &apigatewayv1alpha1.Service{
+			Name:       &name,
+			Port:       &port,
+			Host:       &host,
+			IsExternal: &isExternal,
+		}
+	}
+}
+
+apiRule = NewAPIRule(
+	subscriptionWithoutWebhookAuth,
+	WithPath(),
+	WithServiceAfter(svcName, host), // instead pass it as an argument to NewAPIRule
+)
+```
+</details>
+
+**Example 3**: Side effect as new function
+
+<details>
+  <summary>Don't</summary>
+
+`WithStatusReady` is not used inside a function which is accepting a functional option as input.
+As a reader you would not expect that it performs a `side-effect` on the `apiRuleNew`.
+
+```go
+// source: https://github.com/kyma-project/kyma/blob/a84f76f674babfc63e369bbba44de76499cf475d/components/eventing-controller/controllers/subscription/beb/reconciler_test.go#L592
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+WithStatusReady(apiRuleNew) // <= WithStatusReady not used inside a function which is accepting a functional option as input (e.g. `APIRuleOption`)
+```
+
+</details>
+
+<details>
+  <summary>Do</summary>
+
+In order to make the side-effect more obvious this example uses a function named `MarkReady` instead.
+
+```go
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+MarkReady(apiRuleNew) // instead consider using another function which explicitly states from the name that it applies a side effect
+```
+</details>
+
+## Testing Pyramid in Kyma
+
+The following describes the different testing levels for Kyma. These levels are usually seen as a [pyramid](https://en.wikipedia.org/wiki/Test_automation#Testing_at_different_levels):
+- at the `lower end` of the pyramid, you should have the `most` test cases because they are usually very `fast`, however, you are limited in what you can test because external dependencies are `mocked`. Unit tests are an example of this.
+- at the `higher` end of the pyramid, there are usually `less` tests because these tests take `longer`. However, these tests are a more realistic scenario of how the end-user will use our product. E2E tests are an example of this.
+
+### Controller Integration Tests (kubebuilder)
+
+**Test Setup**
+Controllers using the kubebuilder framework bootstrap integration tests using `Ginkgo` (for testing) and `Gomega` (for assertions).
+Using the `controller-runtime/pkg/envtest` package, a kubernetes `control-plane` (`API server` and `etcd`) is started locally once before the first test runs (`BeforeSuite`).
+The option to use a custom kubernetes cluster exists as well (see `USE_EXISTING_CLUSTER` environment variable).
```

Comment:
I think your first suggestions makes more sense here because the intention is not to provide an alternative. It is more like an additional information.